[ Skip to content ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_aimodelsfallback>)
[ ![logo](https://ai.pydantic.dev/img/logo-white.svg) ](https://ai.pydantic.dev/api/models/fallback/..> "PydanticAI")
PydanticAI 
pydantic_ai.models.fallback 
Type to start searching
[ pydantic/pydantic-ai  ](https://ai.pydantic.dev/api/models/fallback/<https:/github.com/pydantic/pydantic-ai> "Go to repository")
[ ![logo](https://ai.pydantic.dev/img/logo-white.svg) ](https://ai.pydantic.dev/api/models/fallback/..> "PydanticAI") PydanticAI 
[ pydantic/pydantic-ai  ](https://ai.pydantic.dev/api/models/fallback/<https:/github.com/pydantic/pydantic-ai> "Go to repository")
  * [ Introduction  ](https://ai.pydantic.dev/api/models/fallback/..>)
  * [ Installation  ](https://ai.pydantic.dev/api/models/install/>)
  * [ Getting Help  ](https://ai.pydantic.dev/api/models/help/>)
  * [ Contributing  ](https://ai.pydantic.dev/api/models/contributing/>)
  * [ Troubleshooting  ](https://ai.pydantic.dev/api/models/troubleshooting/>)
  * Documentation  Documentation 
    * [ Agents  ](https://ai.pydantic.dev/api/models/agents/>)
    * [ Models  ](https://ai.pydantic.dev/api/models/models/>)
    * [ Dependencies  ](https://ai.pydantic.dev/api/models/dependencies/>)
    * [ Function Tools  ](https://ai.pydantic.dev/api/models/tools/>)
    * [ Common Tools  ](https://ai.pydantic.dev/api/models/common_tools/>)
    * [ Results  ](https://ai.pydantic.dev/api/models/results/>)
    * [ Messages and chat history  ](https://ai.pydantic.dev/api/models/message-history/>)
    * [ Testing and Evals  ](https://ai.pydantic.dev/api/models/testing-evals/>)
    * [ Debugging and Monitoring  ](https://ai.pydantic.dev/api/models/logfire/>)
    * [ Multi-agent Applications  ](https://ai.pydantic.dev/api/models/multi-agent-applications/>)
    * [ Graphs  ](https://ai.pydantic.dev/api/models/graph/>)
    * [ Image and Audio Input  ](https://ai.pydantic.dev/api/models/input/>)
  * [ Examples  ](https://ai.pydantic.dev/api/models/examples/>)
Examples 
    * [ Pydantic Model  ](https://ai.pydantic.dev/api/models/examples/pydantic-model/>)
    * [ Weather agent  ](https://ai.pydantic.dev/api/models/examples/weather-agent/>)
    * [ Bank support  ](https://ai.pydantic.dev/api/models/examples/bank-support/>)
    * [ SQL Generation  ](https://ai.pydantic.dev/api/models/examples/sql-gen/>)
    * [ Flight booking  ](https://ai.pydantic.dev/api/models/examples/flight-booking/>)
    * [ RAG  ](https://ai.pydantic.dev/api/models/examples/rag/>)
    * [ Stream markdown  ](https://ai.pydantic.dev/api/models/examples/stream-markdown/>)
    * [ Stream whales  ](https://ai.pydantic.dev/api/models/examples/stream-whales/>)
    * [ Chat App with FastAPI  ](https://ai.pydantic.dev/api/models/examples/chat-app/>)
    * [ Question Graph  ](https://ai.pydantic.dev/api/models/examples/question-graph/>)
  * API Reference  API Reference 
    * [ pydantic_ai.agent  ](https://ai.pydantic.dev/api/models/fallback/agent/>)
    * [ pydantic_ai.tools  ](https://ai.pydantic.dev/api/models/fallback/tools/>)
    * [ pydantic_ai.common_tools  ](https://ai.pydantic.dev/api/models/fallback/common_tools/>)
    * [ pydantic_ai.result  ](https://ai.pydantic.dev/api/models/fallback/result/>)
    * [ pydantic_ai.messages  ](https://ai.pydantic.dev/api/models/fallback/messages/>)
    * [ pydantic_ai.exceptions  ](https://ai.pydantic.dev/api/models/fallback/exceptions/>)
    * [ pydantic_ai.settings  ](https://ai.pydantic.dev/api/models/fallback/settings/>)
    * [ pydantic_ai.usage  ](https://ai.pydantic.dev/api/models/fallback/usage/>)
    * [ pydantic_ai.format_as_xml  ](https://ai.pydantic.dev/api/models/fallback/format_as_xml/>)
    * [ pydantic_ai.models  ](https://ai.pydantic.dev/api/models/fallback/<../base/>)
    * [ pydantic_ai.models.openai  ](https://ai.pydantic.dev/api/models/fallback/<../openai/>)
    * [ pydantic_ai.models.anthropic  ](https://ai.pydantic.dev/api/models/fallback/<../anthropic/>)
    * [ pydantic_ai.models.cohere  ](https://ai.pydantic.dev/api/models/fallback/<../cohere/>)
    * [ pydantic_ai.models.gemini  ](https://ai.pydantic.dev/api/models/fallback/<../gemini/>)
    * [ pydantic_ai.models.vertexai  ](https://ai.pydantic.dev/api/models/fallback/<../vertexai/>)
    * [ pydantic_ai.models.groq  ](https://ai.pydantic.dev/api/models/fallback/<../groq/>)
    * [ pydantic_ai.models.mistral  ](https://ai.pydantic.dev/api/models/fallback/<../mistral/>)
    * [ pydantic_ai.models.test  ](https://ai.pydantic.dev/api/models/fallback/<../test/>)
    * [ pydantic_ai.models.function  ](https://ai.pydantic.dev/api/models/fallback/<../function/>)
    * pydantic_ai.models.fallback  [ pydantic_ai.models.fallback  ](https://ai.pydantic.dev/api/models/fallback/<./>) Table of contents 
      * [ fallback  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback>)
      * [ FallbackModel  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel>)
        * [ __init__  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel.__init__>)
        * [ request  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel.request>)
        * [ request_stream  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel.request_stream>)
        * [ model_name  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel.model_name>)
        * [ system  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel.system>)
    * [ pydantic_graph  ](https://ai.pydantic.dev/api/models/fallback/pydantic_graph/graph/>)
    * [ pydantic_graph.nodes  ](https://ai.pydantic.dev/api/models/fallback/pydantic_graph/nodes/>)
    * [ pydantic_graph.state  ](https://ai.pydantic.dev/api/models/fallback/pydantic_graph/state/>)
    * [ pydantic_graph.mermaid  ](https://ai.pydantic.dev/api/models/fallback/pydantic_graph/mermaid/>)
    * [ pydantic_graph.exceptions  ](https://ai.pydantic.dev/api/models/fallback/pydantic_graph/exceptions/>)


Table of contents 
  * [ fallback  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback>)
  * [ FallbackModel  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel>)
    * [ __init__  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel.__init__>)
    * [ request  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel.request>)
    * [ request_stream  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel.request_stream>)
    * [ model_name  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel.model_name>)
    * [ system  ](https://ai.pydantic.dev/api/models/fallback/<#pydantic_ai.models.fallback.FallbackModel.system>)


  1. [ Introduction  ](https://ai.pydantic.dev/api/models/fallback/..>)
  2. [ API Reference  ](https://ai.pydantic.dev/api/models/fallback/agent/>)


Version
Showing documentation for the latest release [v0.0.31 2025-03-03](https://ai.pydantic.dev/api/models/fallback/<https:/github.com/pydantic/pydantic-ai/releases/tag/v0.0.31>).
# pydantic_ai.models.fallback
###  FallbackModel `dataclass`
Bases: `Model[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.Model> "pydantic_ai.models.Model")`
A model that uses one or more fallback models upon failure.
Apart from `__init__`, all methods are private or match those of the base class.
Source code in `pydantic_ai_slim/pydantic_ai/models/fallback.py`
```
 17
 18
 19
 20
 21
 22
 23
 24
 25
 26
 27
 28
 29
 30
 31
 32
 33
 34
 35
 36
 37
 38
 39
 40
 41
 42
 43
 44
 45
 46
 47
 48
 49
 50
 51
 52
 53
 54
 55
 56
 57
 58
 59
 60
 61
 62
 63
 64
 65
 66
 67
 68
 69
 70
 71
 72
 73
 74
 75
 76
 77
 78
 79
 80
 81
 82
 83
 84
 85
 86
 87
 88
 89
 90
 91
 92
 93
 94
 95
 96
 97
 98
 99
100
101
102
103
104
105
106
107
```
| ```
@dataclass(init=False)
classFallbackModel(Model):
"""A model that uses one or more fallback models upon failure.
  Apart from `__init__`, all methods are private or match those of the base class.
  """
  models: list[Model]
  _model_name: str = field(repr=False)
  _fallback_on: Callable[[Exception], bool]
  def__init__(
    self,
    default_model: Model | KnownModelName,
    *fallback_models: Model | KnownModelName,
    fallback_on: Callable[[Exception], bool] | tuple[type[Exception], ...] = (ModelHTTPError,),
  ):
"""Initialize a fallback model instance.
    Args:
      default_model: The name or instance of the default model to use.
      fallback_models: The names or instances of the fallback models to use upon failure.
      fallback_on: A callable or tuple of exceptions that should trigger a fallback.
    """
    self.models = [infer_model(default_model), *[infer_model(m) for m in fallback_models]]
    self._model_name = f'FallBackModel[{", ".join(model.model_nameformodelinself.models)}]'
    if isinstance(fallback_on, tuple):
      self._fallback_on = _default_fallback_condition_factory(fallback_on)
    else:
      self._fallback_on = fallback_on
  async defrequest(
    self,
    messages: list[ModelMessage],
    model_settings: ModelSettings | None,
    model_request_parameters: ModelRequestParameters,
  ) -> tuple[ModelResponse, Usage]:
"""Try each model in sequence until one succeeds.
    In case of failure, raise a FallbackExceptionGroup with all exceptions.
    """
    exceptions: list[Exception] = []
    for model in self.models:
      try:
        return await model.request(messages, model_settings, model_request_parameters)
      except Exception as exc:
        if self._fallback_on(exc):
          exceptions.append(exc)
          continue
        raise exc
    raise FallbackExceptionGroup('All models from FallbackModel failed', exceptions)
  @asynccontextmanager
  async defrequest_stream(
    self,
    messages: list[ModelMessage],
    model_settings: ModelSettings | None,
    model_request_parameters: ModelRequestParameters,
  ) -> AsyncIterator[StreamedResponse]:
"""Try each model in sequence until one succeeds."""
    exceptions: list[Exception] = []
    for model in self.models:
      async with AsyncExitStack() as stack:
        try:
          response = await stack.enter_async_context(
            model.request_stream(messages, model_settings, model_request_parameters)
          )
        except Exception as exc:
          if self._fallback_on(exc):
            exceptions.append(exc)
            continue
          raise exc
        yield response
        return
    raise FallbackExceptionGroup('All models from FallbackModel failed', exceptions)
  @property
  defmodel_name(self) -> str:
"""The model name."""
    return self._model_name
  @property
  defsystem(self) -> str | None:
"""The system / model provider, n/a for fallback models."""
    return None

```
  
---|---  
####  __init__
```
__init__(
  default_model: Model[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.Model> "pydantic_ai.models.Model") | KnownModelName[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.KnownModelName> "pydantic_ai.models.KnownModelName"),
  *fallback_models: Model[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.Model> "pydantic_ai.models.Model") | KnownModelName[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.KnownModelName> "pydantic_ai.models.KnownModelName"),
  fallback_on: (
    Callable[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/typing.html#typing.Callable> "typing.Callable")[[Exception[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/exceptions.html#Exception>)], bool[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/functions.html#bool>)]
    | tuple[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/stdtypes.html#tuple>)[type[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/functions.html#type>)[Exception[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/exceptions.html#Exception>)], ...]
  ) = (ModelHTTPError[](https://ai.pydantic.dev/api/models/fallback/exceptions/#pydantic_ai.exceptions.ModelHTTPError> "pydantic_ai.exceptions.ModelHTTPError"),)
)

```

Initialize a fallback model instance.
Parameters:
Name | Type | Description | Default  
---|---|---|---  
`default_model` |  `Model[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.Model> "pydantic_ai.models.Model") | KnownModelName[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.KnownModelName> "pydantic_ai.models.KnownModelName")` |  The name or instance of the default model to use. |  _required_  
`fallback_models` |  `Model[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.Model> "pydantic_ai.models.Model") | KnownModelName[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.KnownModelName> "pydantic_ai.models.KnownModelName")` |  The names or instances of the fallback models to use upon failure. |  `()`  
`fallback_on` |  `Callable[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/typing.html#typing.Callable> "typing.Callable")[[Exception[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/exceptions.html#Exception>)], bool[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/functions.html#bool>)] | tuple[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/stdtypes.html#tuple>)[type[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/functions.html#type>)[Exception[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/exceptions.html#Exception>)], ...]` |  A callable or tuple of exceptions that should trigger a fallback. |  `(ModelHTTPError[](https://ai.pydantic.dev/api/models/fallback/exceptions/#pydantic_ai.exceptions.ModelHTTPError> "pydantic_ai.exceptions.ModelHTTPError"),)`  
Source code in `pydantic_ai_slim/pydantic_ai/models/fallback.py`
```
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
```
| ```
def__init__(
  self,
  default_model: Model | KnownModelName,
  *fallback_models: Model | KnownModelName,
  fallback_on: Callable[[Exception], bool] | tuple[type[Exception], ...] = (ModelHTTPError,),
):
"""Initialize a fallback model instance.
  Args:
    default_model: The name or instance of the default model to use.
    fallback_models: The names or instances of the fallback models to use upon failure.
    fallback_on: A callable or tuple of exceptions that should trigger a fallback.
  """
  self.models = [infer_model(default_model), *[infer_model(m) for m in fallback_models]]
  self._model_name = f'FallBackModel[{", ".join(model.model_nameformodelinself.models)}]'
  if isinstance(fallback_on, tuple):
    self._fallback_on = _default_fallback_condition_factory(fallback_on)
  else:
    self._fallback_on = fallback_on

```
  
---|---  
####  request `async`
```
request(
  messages: list[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/stdtypes.html#list>)[ModelMessage[](https://ai.pydantic.dev/api/models/fallback/messages/#pydantic_ai.messages.ModelMessage> "pydantic_ai.messages.ModelMessage")],
  model_settings: ModelSettings[](https://ai.pydantic.dev/api/models/fallback/settings/#pydantic_ai.settings.ModelSettings> "pydantic_ai.settings.ModelSettings") | None,
  model_request_parameters: ModelRequestParameters[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.ModelRequestParameters> "pydantic_ai.models.ModelRequestParameters"),
) -> tuple[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/stdtypes.html#tuple>)[ModelResponse[](https://ai.pydantic.dev/api/models/fallback/messages/#pydantic_ai.messages.ModelResponse> "pydantic_ai.messages.ModelResponse"), Usage[](https://ai.pydantic.dev/api/models/fallback/usage/#pydantic_ai.usage.Usage> "pydantic_ai.usage.Usage")]

```

Try each model in sequence until one succeeds.
In case of failure, raise a FallbackExceptionGroup with all exceptions.
Source code in `pydantic_ai_slim/pydantic_ai/models/fallback.py`
```
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
```
| ```
async defrequest(
  self,
  messages: list[ModelMessage],
  model_settings: ModelSettings | None,
  model_request_parameters: ModelRequestParameters,
) -> tuple[ModelResponse, Usage]:
"""Try each model in sequence until one succeeds.
  In case of failure, raise a FallbackExceptionGroup with all exceptions.
  """
  exceptions: list[Exception] = []
  for model in self.models:
    try:
      return await model.request(messages, model_settings, model_request_parameters)
    except Exception as exc:
      if self._fallback_on(exc):
        exceptions.append(exc)
        continue
      raise exc
  raise FallbackExceptionGroup('All models from FallbackModel failed', exceptions)

```
  
---|---  
####  request_stream `async`
```
request_stream(
  messages: list[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/stdtypes.html#list>)[ModelMessage[](https://ai.pydantic.dev/api/models/fallback/messages/#pydantic_ai.messages.ModelMessage> "pydantic_ai.messages.ModelMessage")],
  model_settings: ModelSettings[](https://ai.pydantic.dev/api/models/fallback/settings/#pydantic_ai.settings.ModelSettings> "pydantic_ai.settings.ModelSettings") | None,
  model_request_parameters: ModelRequestParameters[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.ModelRequestParameters> "pydantic_ai.models.ModelRequestParameters"),
) -> AsyncIterator[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/collections.abc.html#collections.abc.AsyncIterator> "collections.abc.AsyncIterator")[StreamedResponse[](https://ai.pydantic.dev/api/models/fallback/<../base/#pydantic_ai.models.StreamedResponse> "pydantic_ai.models.StreamedResponse")]

```

Try each model in sequence until one succeeds.
Source code in `pydantic_ai_slim/pydantic_ai/models/fallback.py`
```
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
```
| ```
@asynccontextmanager
async defrequest_stream(
  self,
  messages: list[ModelMessage],
  model_settings: ModelSettings | None,
  model_request_parameters: ModelRequestParameters,
) -> AsyncIterator[StreamedResponse]:
"""Try each model in sequence until one succeeds."""
  exceptions: list[Exception] = []
  for model in self.models:
    async with AsyncExitStack() as stack:
      try:
        response = await stack.enter_async_context(
          model.request_stream(messages, model_settings, model_request_parameters)
        )
      except Exception as exc:
        if self._fallback_on(exc):
          exceptions.append(exc)
          continue
        raise exc
      yield response
      return
  raise FallbackExceptionGroup('All models from FallbackModel failed', exceptions)

```
  
---|---  
####  model_name `property`
```
model_name: str[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/stdtypes.html#str>)

```

The model name.
####  system `property`
```
system: str[](https://ai.pydantic.dev/api/models/fallback/<https:/docs.python.org/3/library/stdtypes.html#str>) | None

```

The system / model provider, n/a for fallback models.
© Pydantic Services Inc. 2024 to present 
